lib/pos_snapshot_cache.go

```diff
@@ -0,0 +1,240 @@
+	leaderScheduleRWLock                  sync.RWMutex
+
+	// snapshot global params cache
+	cachedGlobalParamsBySnapshotAtEpoch map[uint64]*GlobalParamsEntry
+	globalParamsRWLock                  sync.RWMutex
+
+	snapshotsCached map[uint64]bool
+}
+
+func NewSnapshotCache() *SnapshotCache {
+	return &SnapshotCache{
+		cachedValidatorsBySnapshotAtEpoch:     make(map[uint64][]*ValidatorEntry),
+		cachedLeaderScheduleBySnapshotAtEpoch: make(map[uint64][]*PKID),
+		cachedGlobalParamsBySnapshotAtEpoch:   make(map[uint64]*GlobalParamsEntry),
+		snapshotsCached:                       make(map[uint64]bool),
+	}
+}
+
+// LoadCacheAtSnapshotAtEpochNumber loads the cache with the snapshot data at the
+// given snapshot epoch number and the previous two snapshot at epochs. It should
+// only ever be called when committing a block.
+func (sc *SnapshotCache) LoadCacheAtSnapshotAtEpochNumber(
+	snapshotAtEpochNumber uint64,
+	currentEpochNumber uint64,
+	handle *badger.DB,
+	snapshot *Snapshot,
+	params *DeSoParams,
+) {
+
+	snapshotsToLoad := []uint64{}
+	for ii := snapshotAtEpochNumber; ii < currentEpochNumber; ii++ {
+		snapshotsToLoad = append(snapshotsToLoad, ii)
+	}
+	for _, epochNumber := range snapshotsToLoad {
+		_, err1 := sc.GetSnapshotValidatorSetEntriesByStakeAtEpochNumber(snapshotAtEpochNumber, handle, snapshot, params)
+		_, err2 := sc.GetSnapshotLeaderSchedule(snapshotAtEpochNumber, handle)
+		_, err3 := sc.GetSnapshotGlobalParams(snapshotAtEpochNumber, handle, snapshot, params)
+		if err1 == nil && err2 == nil && err3 == nil {
+			sc.snapshotsCached[epochNumber] = true
+		}
+	}
+	for epochNumber := range sc.snapshotsCached {
+		if epochNumber < snapshotAtEpochNumber {
+			delete(sc.cachedValidatorsBySnapshotAtEpoch, epochNumber)
+			delete(sc.cachedLeaderScheduleBySnapshotAtEpoch, epochNumber)
+			delete(sc.cachedGlobalParamsBySnapshotAtEpoch, epochNumber)
+			delete(sc.snapshotsCached, epochNumber)
+		}
+	}
+}
+
+// GetAllCachedSnapshotValidatorSetEntries returns all cached validator entries.
+func (sc *SnapshotCache) GetAllCachedSnapshotValidatorSetEntries() (map[uint64][]*ValidatorEntry, bool) {
+	if sc == nil {
+		return nil, false
+	}
+	sc.validatorRWLock.RLock()
+	defer sc.validatorRWLock.RUnlock()
+	return sc.cachedValidatorsBySnapshotAtEpoch, true
+}
+
+// GetSnapshotValidatorSetEntriesByStakeAtEpochNumber returns the top N validators by stake at the given snapshot
+// epoch number.
+func (sc *SnapshotCache) GetSnapshotValidatorSetEntriesByStakeAtEpochNumber(
+	snapshotAtEpochNumber uint64,
+	handle *badger.DB,
+	snapshot *Snapshot,
+	params *DeSoParams,
+) ([]*ValidatorEntry, error) {
+	// First check the cache to see if we have the validator entries for the snapshot epoch number.
+	if sc != nil {
+		sc.validatorRWLock.RLock()
+		validatorEntries, exists := sc.cachedValidatorsBySnapshotAtEpoch[snapshotAtEpochNumber]
+		sc.validatorRWLock.RUnlock()
+		if exists {
+			return validatorEntries, nil
+		}
+	}
```

Comment:
If the cache is nil, this should error. It's up to the caller to know whether they have a nil pointer vs a fully constructed cache
```suggestion
	// First check the cache to see if we have the validator entries for the snapshot epoch number.
	sc.validatorRWLock.RLock()
	validatorEntries, exists := sc.cachedValidatorsBySnapshotAtEpoch[snapshotAtEpochNumber]
	sc.validatorRWLock.RUnlock()
	if exists {
		return validatorEntries, nil
	}
```